File: OneOffs/gym/102367/C/C.cpp
// MuMonash

#include <iostream>

using namespace std;

typedef pair<int, int> pii;

#define EMPTY 0
#define KING 1
#define ENEMY 2
#define SPOTTED 3

int main() {

    int n;
    cin >> n >> ws;

    pii king_position;

    int BOARD[n][n];
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            char x;
            cin >> x;
            if (x == '*') {
                BOARD[i][j] = ENEMY;
            } else if (x == 'K') {
                BOARD[i][j] = KING;
                king_position = pii(i, j);
            } else {
                BOARD[i][j] = EMPTY;
            }
        }
    }

    int pawns = 0;
    bool possible = true;

    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            if (BOARD[i][j] == ENEMY) {
                if (abs(i - king_position.first) <= 1 && abs(j - king_position.second) <= 1) {
                    BOARD[i][j] = SPOTTED;
                    continue;
                }
                if (i+1 < n && j+1 < n && BOARD[i+1][j+1] == EMPTY) {
                    // Choose the right side where possible
                    pawns += 1;
                    if (j+2 < n && BOARD[i][j+2] == ENEMY) BOARD[i][j+2] = SPOTTED;
                } else if (i+1 < n && j-1 < n && BOARD[i+1][j-1] == EMPTY) {
                    pawns += 1;
                } else {
                    possible = false;
                }
            }
        }
    }

    if (possible) {
        cout << pawns << endl;
    } else {
        cout << -1 << endl;
    }

    return 0;
}

File: OneOffs/gym/102367/C/C.py
# MuMonash
# THIS TLE?????

n = int(input())
board = [
    [x for x in input()]
    for _ in range(n)
]

king_location = None
bad_pieces = []
for x in range(n):
    for y in range(n):
        if board[x][y] == 'K':
            king_location = (x, y)
        elif board[x][y] == '*':
            bad_pieces.append((x, y))

num_pawns = 0
bad = False

for x, y in bad_pieces:
    if abs(x-king_location[0]) <= 1 and abs(y-king_location[1]) <= 1:
        continue
    if board[x][y] == 'x':
        continue
    if 0 <= x+1 < n and 0 <= y+1 < n and board[x+1][y+1] == '-':
        # Always place it on the right side when possible.
        board[x][y] = 'x'
        board[x+1][y+1] = 'p'
        num_pawns += 1
        if y+2 < n and board[x][y+2] == '*':
            board[x][y+2] = 'x'
    elif 0 <= x+1 < n and 0 <= y-1 < n and board[x+1][y-1] == '-':
        # Otherwise the left will do.
        board[x][y] = 'x'
        board[x+1][y-1] = 'p'
        num_pawns += 1
    else:
        bad = True

if bad:
    print(-1)
else:
    print(num_pawns)
